### PR TITLE
feat: Topology aware prompt MONGOSH-196

### PR DESCRIPTION
**Pull request by rose-m**
This adds the topology aware prompt but in such a way that we can easily extend that to support a user-customizable prompt in the same way the old shell did.

**Connecting to an Atlas Cluster:**

image

**Connecting to a local node with replica set, not initialized, then initializing:**

image

**Compass Browser Repl:**

image

### CODE REVIEW

File: packages/cli-repl/src/mongosh-repl.ts
Comment (addaleax) [inline packages/cli-repl/src/mongosh-repl.ts:242]: We could also move this into the `finally`, right?

Comment (rose-m) [reply packages/cli-repl/src/mongosh-repl.ts:242]: Yep, true... will move it.

File: packages/shell-api/src/shell-internal-state.ts
Comment (addaleax) [inline packages/shell-api/src/shell-internal-state.ts:321]: Do we want to keep the uppercase names? That has a very 80s feel to it for me … :)

Comment (rose-m) [reply packages/shell-api/src/shell-internal-state.ts:321]: :D I'm fine with either casing... would you suggest just having it lowercase?

Comment (addaleax) [reply packages/shell-api/src/shell-internal-state.ts:320]: I think either lowercase or capitalized would be a bit nicer, but I don’t have a strong preference anyway :)

**Comment by aherlihy** [comment]
LGTM! Can you remind me if you looked into using the SDAM monitoring to determine the prompt? I remember chatting briefly about it during standup but I don't remember if you did a spike or it was determined to be too much work for the beta. The benefit of using SDAM would be the driver would notify us if there was a topology change, so we would be able to automatically update the prompt to match the topology, and we would not have to potentially run an extra command for every single line of user input.

**Comment by rose-m** [comment]
@aherlihy as we discussed yesterday, we could move the default prompt also in the `ServiceProvider` and then rely on the topology of the driver instead of running additional commands. However, right now their types are not properly exported - or rather only the types and not certain enums - which would make it pretty fragile to implement currently. I'd rather suggest merging what we already have and once the driver fixes their exports go on and move it to the `ServiceProvider`.

Of course we can also just put this on hold until it is resolved.

See: https://mongodb.slack.com/archives/C0V8RU15L/p1609926543248900

File: packages/browser-repl/src/components/shell-input.tsx
Comment (addaleax) [inline packages/browser-repl/src/components/shell-input.tsx:122]: Nice trick :)

File: packages/cli-repl/src/mongosh-repl.ts
Comment (addaleax) [inline packages/cli-repl/src/mongosh-repl.ts:299]: I don’t think we need to care about the prompt for programmatic usage, so something like this might be nice for saving some DB calls when we’re running a script:

```suggestion
      if (connectionIssues || !this.input.isTTY) {
```

File: packages/cli-repl/test/test-shell.ts
Comment (addaleax) [inline packages/cli-repl/test/test-shell.ts:121]: There should be no `>` characters left after the `split('>')`, so I think that last `.replace()` can be dropped?

**Review by addaleax — APPROVED** [review]

File: packages/browser-runtime-core/src/open-context-runtime.ts
Comment (addaleax) [inline packages/browser-runtime-core/src/open-context-runtime.ts:69]: Hm… maybe this could still forward to something like `this.internalState.getShellPrompt()`, which then forwards to the service provider? That way we can avoid storing it here just for this

File: packages/service-provider-server/src/cli-service-provider.ts
Comment (addaleax) [inline packages/service-provider-server/src/cli-service-provider.ts:260]: Tbh, I think using the strings as-is is just fine – I had understood the problem as us having to do some sort of casting, but since TypeScript would complain about invalid values in the `switch` I think we’re good here

**Review by addaleax — APPROVED** [review]

**Comment by addaleax** [comment]
Btw, `ShellInternalState` seemed like the right place to implement this… we’re already allowing `getTopology()` to return `undefined`, so there’s no additional requirements to service providers in general, and I don’t think relying on driver internals is something we need to worry about since this is all public. If you want to make this a bit easier for the shell, one thing that could be nice (also for #532), would be processing the topology and server descriptions in the service provider, and then returning another enum value based on that to the shell-api package?

File: packages/browser-runtime-core/src/open-context-runtime.ts
Comment (rose-m) [reply packages/browser-runtime-core/src/open-context-runtime.ts:69]: Yeah I cleaned up the remnants of the `internalState` now. All places now go to the `serviceProvider` directly. The forwarding didn't really make sense to me anymore since also the user-specific prompt will be implemented and evaluated in `MongoshNodeRepl` which basically has the `serviceProvider` directly already, too. So it made sense to me to get rid of the unnecessary level of indirection.

File: packages/service-provider-server/src/cli-service-provider.ts
Comment (rose-m) [reply packages/service-provider-server/src/cli-service-provider.ts:260]: Yeah indeed - in the switch it works but it complained in the tests when I tried to use it typed. I'd leave it in there just to be sure to make it consistent.

File: packages/browser-runtime-core/src/open-context-runtime.ts
Comment (addaleax) [reply packages/browser-runtime-core/src/open-context-runtime.ts:69]: I’m just a bit concerned because this now means that this package has to have knowledge about what the service providers look like, rather than it being just something that it passes through to the `ShellInternalState` constructor

Comment (rose-m) [reply packages/browser-runtime-core/src/open-context-runtime.ts:69]: Ah very valid point... mh... I've no hard feelings on that one. True, the runtimes just forward that to the internal state...

**Review by addaleax — APPROVED** [review]

**Pull request merged**
